lib/spack/spack/util/path.py

```diff
@@ -51,27 +53,71 @@ def get_user():
         return getpass.getuser()
 
 
-# Substitutions to perform
-def replacements():
-    # break circular import from spack.util.executable
+#: dictionary of replacements for path substitutions
+Replacer = Callable[[str], str]
+_replacements: Dict[str, Replacer] = {}
+
+
+class replacer:
+    """Decorator to put replacer functions in ``_replacements`` dict."""
+
+    def __init__(self, *keys):
+        self.keys: List[str] = keys
+
+    def __call__(self, function: Replacer):
+        for key in self.keys:
+            _replacements[key] = function
+        return function
+
+
+@replacer("spack", "user_cache_path")
+def _replace_path_key(key):
     import spack.paths
 
-    arch = architecture()
+    return getattr(spack.paths, key)
+
+
+@replacer("user")
+def _replace_user(key):
+    return get_user()
+
+
+@replacer("tempdir")
+def _replace_tempdir(key):
+    return tempfile.gettempdir()
+
+
+@replacer("arch", "architecture")
+def _replace_arch(key):
+    return str(architecture())
+
 
-    return {
-        "spack": spack.paths.prefix,
-        "user": get_user(),
-        "tempdir": tempfile.gettempdir(),
-        "user_cache_path": spack.paths.user_cache_path,
-        "architecture": str(arch),
-        "arch": str(arch),
-        "platform": str(arch.platform),
-        "operating_system": str(arch.os),
-        "os": str(arch.os),
-        "target": str(arch.target),
-        "target_family": str(arch.target.microarchitecture.family),
-        "date": date.today().strftime("%Y-%m-%d"),
-    }
+@replacer("os", "operating_system")
+def _replace_os(key):
+    return str(architecture().os)
+
+
+@replacer("platform", "target")
+def _replace_arch_attr(key):
+    return str(getattr(architecture(), key))
```

Comment:
This is heinous to read, I was convinced it was a bug until I was about to post the comment saying so. At least it needs to be explained in comments.